qualtran/bloqs/arithmetic.py

```diff
@@ -176,6 +234,152 @@ def t_complexity(self):
+        return "r*i"
+
+    def t_complexity(self):
+        # Eq. D8, we are assuming dA and dB there are assumed as inputs and the
+        # user has ensured these are large enough for their desired precision.
+        num_toff = self.r_bitsize * (2 * self.i_bitsize - 1) - self.i_bitsize**2
+        return TComplexity(t=4 * num_toff)
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set[Tuple[int, Bloq]]:
+        # Eq. D8, we are assuming dA(r_bitsize) and dB(i_bitsize) are inputs and
+        # the user has ensured these are large enough for their desired
+        # precision.
+        num_toff = self.r_bitsize * (2 * self.i_bitsize - 1) - self.i_bitsize**2
+        return {(4 * num_toff, TGate())}
+
+
+@frozen
+class MultiplyTwoReals(Bloq):
+    r"""Multiply two fixed-point representations of real numbers
+
+    i.e.
+
+    $$
+        |a\rangle|b\rangle|0\rangle \rightarrow |a\rangle|b\rangle|a \times b\rangle
+    $$
+
+    The real numbers are assumed to be in the range [0, 1).
+
+    Args:
+        bitsize: Number of bits used to represent the real number.
+
+    Registers:
+     - a: bitsize-sized input register.
+     - b: bitsize-sized input register.
+     - result: bitsize output register
+
+    References:
+        [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization
+            ](https://arxiv.org/pdf/2007.07391.pdf) pg 71.
+    """
+
+    bitsize: int
+
+    @property
+    def signature(self):
+        return Signature.build(a=self.bitsize, b=self.bitsize, result=self.bitsize)
+
+    def pretty_name(self) -> str:
+        return "a*b"
+
+    def t_complexity(self):
+        # Eq. D13, there it is suggested keeping both registers the same size is optimal.
+        num_toff = self.bitsize**2 - self.bitsize - 1
+        return TComplexity(t=4 * num_toff)
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set[Tuple[int, Bloq]]:
+        # Eq. D13, there it is suggested keeping both registers the same size is optimal.
+        num_toff = self.bitsize**2 - self.bitsize - 1
+        return {(4 * num_toff, TGate())}
+
+
+@frozen
+class SquareRealNumber(Bloq):
+    r"""Square a fixed-point representation of a real number
+
+        i.e.
+
+        $$
+            |a\rangle|0\rangle \rightarrow |a\rangle||a^2\rangle
+        $$
+
+        The real numbers are assumed to be in the range [0, 1).
+
+        Args:
+            bitsize: Number of bits used to represent the real number.
+
+        Registers:
+         - a: bitsize-sized input register.
+         - b: bitsize-sized input register.
+         - result: bitsize output register
+
+        References:
+            [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization
+    ](https://arxiv.org/pdf/2007.07391.pdf) pg 74.
+    """
+
+    bitsize: int
+
+    def __attrs_post_init__(self):
+        assert self.bitsize > 3, "Must have at least 3 bits for bloq to make sense."
```

Comment:
generally, try to use an `if not.. : raise ValueError()` and here it wouldn't hurt to say what the bloq is :)

Reply:
Done.